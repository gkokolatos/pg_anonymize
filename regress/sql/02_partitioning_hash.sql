LOAD 'pg_anonymize';
-- mask our own user
SELECT current_user \gset
SECURITY LABEL FOR pg_anonymize ON ROLE :current_user IS 'anonymize';

--------------------
-- hash partitioning
--------------------
CREATE TABLE t_part_hash(id integer, val text) PARTITION BY HASH (id);
CREATE TABLE t_part_hash_0 PARTITION OF t_part_hash FOR VALUES WITH (MODULUS 2, REMAINDER 0);
CREATE TABLE t_part_hash_1 PARTITION OF t_part_hash FOR VALUES WITH (MODULUS 2, REMAINDER 1);
INSERT INTO t_part_hash SELECT i, 'line ' || i FROM generate_series(1, 4) i;
SECURITY LABEL FOR pg_anonymize
    ON COLUMN public.t_part_hash.val IS $$'root hidden'::text$$;
SECURITY LABEL FOR pg_anonymize
    ON COLUMN public.t_part_hash_1.val IS $$'part hidden'::text$$;

SET pg_anonymize.enabled = 'off';

-- should see original data
SELECT * FROM t_part_hash ORDER BY id;
COPY t_part_hash TO STDOUT;
SELECT * FROM t_part_hash_0 ORDER BY id;
COPY t_part_hash_0 TO STDOUT;
SELECT * FROM t_part_hash_1 ORDER BY id;
COPY t_part_hash_1 TO STDOUT;

SET pg_anonymize.enabled = 'on';
SET pg_anonymize.inherit_labels = false;

--should see anonymized data when selecting from root partition
SELECT * FROM t_part_hash ORDER BY id;
COPY t_part_hash TO STDOUT;
-- but original data from any leaf partition when label inheritance is disabled
SELECT * FROM t_part_hash_0 ORDER BY id;
COPY t_part_hash_0 TO STDOUT;
-- unless there's an explicit anonymization rule on it
SELECT * FROM t_part_hash_1 ORDER BY id;
COPY t_part_hash_1 TO STDOUT;

SET pg_anonymize.inherit_labels = true;
-- should see anonymized data from inherited security labels
SELECT * FROM t_part_hash_0 ORDER BY id;
COPY t_part_hash_0 TO STDOUT;
-- and still see same anonymized data
SELECT * FROM t_part_hash_1 ORDER BY id;
COPY t_part_hash_1 TO STDOUT;
