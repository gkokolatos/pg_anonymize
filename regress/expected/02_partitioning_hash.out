LOAD 'pg_anonymize';
-- mask our own user
SELECT current_user \gset
SECURITY LABEL FOR pg_anonymize ON ROLE :current_user IS 'anonymize';
--------------------
-- hash partitioning
--------------------
CREATE TABLE t_part_hash(id integer, val text) PARTITION BY HASH (id);
CREATE TABLE t_part_hash_0 PARTITION OF t_part_hash FOR VALUES WITH (MODULUS 2, REMAINDER 0);
CREATE TABLE t_part_hash_1 PARTITION OF t_part_hash FOR VALUES WITH (MODULUS 2, REMAINDER 1);
INSERT INTO t_part_hash SELECT i, 'line ' || i FROM generate_series(1, 4) i;
SECURITY LABEL FOR pg_anonymize
    ON COLUMN public.t_part_hash.val IS $$'root hidden'::text$$;
SECURITY LABEL FOR pg_anonymize
    ON COLUMN public.t_part_hash_1.val IS $$'part hidden'::text$$;
SET pg_anonymize.enabled = 'off';
-- should see original data
SELECT * FROM t_part_hash ORDER BY id;
 id |  val   
----+--------
  1 | line 1
  2 | line 2
  3 | line 3
  4 | line 4
(4 rows)

COPY t_part_hash TO STDOUT;
ERROR:  cannot copy from partitioned table "t_part_hash"
HINT:  Try the COPY (SELECT ...) TO variant.
SELECT * FROM t_part_hash_0 ORDER BY id;
 id |  val   
----+--------
  1 | line 1
  2 | line 2
(2 rows)

COPY t_part_hash_0 TO STDOUT;
1	line 1
2	line 2
SELECT * FROM t_part_hash_1 ORDER BY id;
 id |  val   
----+--------
  3 | line 3
  4 | line 4
(2 rows)

COPY t_part_hash_1 TO STDOUT;
3	line 3
4	line 4
SET pg_anonymize.enabled = 'on';
SET pg_anonymize.inherit_labels = false;
--should see anonymized data when selecting from root partition
SELECT * FROM t_part_hash ORDER BY id;
 id |     val     
----+-------------
  1 | root hidden
  2 | root hidden
  3 | root hidden
  4 | root hidden
(4 rows)

COPY t_part_hash TO STDOUT;
ERROR:  cannot copy from partitioned table "t_part_hash"
HINT:  Try the COPY (SELECT ...) TO variant.
-- but original data from any leaf partition when label inheritance is disabled
SELECT * FROM t_part_hash_0 ORDER BY id;
 id |  val   
----+--------
  1 | line 1
  2 | line 2
(2 rows)

COPY t_part_hash_0 TO STDOUT;
1	line 1
2	line 2
-- unless there's an explicit anonymization rule on it
SELECT * FROM t_part_hash_1 ORDER BY id;
 id |     val     
----+-------------
  3 | part hidden
  4 | part hidden
(2 rows)

COPY t_part_hash_1 TO STDOUT;
3	part hidden
4	part hidden
SET pg_anonymize.inherit_labels = true;
-- should see anonymized data from inherited security labels
SELECT * FROM t_part_hash_0 ORDER BY id;
 id |     val     
----+-------------
  1 | root hidden
  2 | root hidden
(2 rows)

COPY t_part_hash_0 TO STDOUT;
1	root hidden
2	root hidden
-- and still see same anonymized data
SELECT * FROM t_part_hash_1 ORDER BY id;
 id |     val     
----+-------------
  3 | part hidden
  4 | part hidden
(2 rows)

COPY t_part_hash_1 TO STDOUT;
3	part hidden
4	part hidden
